test/distributed/optim/test_zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,511 @@
+                        dist.gather(p, receptacle, dst=0, group=process_group)
+                        if self.rank == 0:
+                            for sync_p in receptacle[1:]:
+                                assert torch.all(torch.eq(receptacle[0], sync_p)), "Models differ in between ranks"
+
+        if self.rank in sub_group_ranks:
+            # Model fitting in the broadcast bucket
+            model = torch.nn.Sequential(torch.nn.Linear(input_width, hidden), torch.nn.Linear(hidden, target_width)).to(
+                device
+            )
+
+            # With SGD, Momentum is required to get a state to shard
+            optimizer = ZeroRedundancyOptimizer(
+                model.parameters(), optim=SGD, lr=0.1, momentum=0.99, group=process_group, bucket_cap_kb=2 ** 10
+            )
+            check(optimizer)
+
+            # Model not-fitting in the broadcast bucket
+            model = torch.nn.Sequential(torch.nn.Linear(input_width, hidden), torch.nn.Linear(hidden, target_width)).to(
+                device
+            )
+
+            # With SGD, Momentum is required to get a state to shard
+            optimizer = ZeroRedundancyOptimizer(
+                model.parameters(), optim=SGD, lr=0.1, momentum=0.99, group=process_group, bucket_cap_kb=0
+            )
+            check(optimizer)
+
+    @skip_if_no_gpu
+    def test_gradient_clipping(self):
+        device = torch.device(self.rank)
+        torch.manual_seed(self.rank)  # make sure that the different rank get different data
+
+        # Run a dummy step so that the optimizer state dict exists
+        batch, input_width, hidden, target_width = 3, 20, 10, 5
+        target = torch.rand((batch, target_width), device=device)
+        inputs = torch.rand((batch, input_width), device=device)
+        NORMS = [1.0, 2.0, 1, 2, inf]
+        CLIP_NORM = 0.3
+
+        def check(norm):
+            model_base = torch.nn.Sequential(
+                torch.nn.Linear(input_width, hidden),
+                torch.nn.Linear(hidden, hidden),
+                torch.nn.Linear(hidden, target_width),
+            ).to(device)
+            model = copy.deepcopy(model_base)
+
+            # For this test the gradients are (all) reduced in the same way in between the torch reference and fairscale.
+            # Normally OSS would use ShardedDDP and only reduce to the proper rank, but this does not change the
+            # gradient norm computation from OSS and adds a dependency.
+            # to keep the comparison apples-to-apples DDP is used in both cases
+            model_oss = DDP(
+                module=model_base,
+                device_ids=[self.rank],
+            )
+            sharded_optimizer = ZeroRedundancyOptimizer(model_oss.parameters(), lr=0.1, momentum=0.99)
+
+            model_ddp = DDP(
+                model,
+                device_ids=[self.rank],
+            )
+
+            loss_fn = torch.nn.L1Loss()
+            loss_fn.to(device)
```

Comment:
Shouldn't we add the following tests: 
1. A test that runs DDP with and without ZeRO optimizer and runs multiple training iterations and verifies the parameters are same for both?
2. Run test 1 with a variety of different optimizers (ex: Adam, Adagrad) instead of only testing SGD.

Reply:
Done on fairscale, I'll update this PR afterwards, I'd like to make sure that the two don't diverge